File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/S2CPlaySoundHandler.java
package generations.gg.generations.core.generationscore.common.network.packets;

import com.cobblemon.mod.common.util.MovingSoundInstance;
import generations.gg.generations.core.generationscore.common.network.ClientNetworkPacketHandler;
import net.minecraft.client.Minecraft;
import net.minecraft.sounds.SoundSource;

public class S2CPlaySoundHandler implements ClientNetworkPacketHandler<S2CPlaySoundPacket> {
    @Override
    public void handle(S2CPlaySoundPacket packet, Minecraft minecraft) {
        if (packet.play()) {
            minecraft.getSoundManager().play(new MovingSoundInstance(packet.soundEvent(), SoundSource.RECORDS, () -> minecraft.player.position(), 1.0f, 1.0f, false, packet.length(), 0));
        } else {
            // Stop playing the sound
            Minecraft.getInstance().getSoundManager().stop(packet.soundEvent().getLocation(), SoundSource.RECORDS);
        }
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/network/packets/S2CPlaySoundHandler.kt
package generations.gg.generations.core.generationscore.common.network.packets

import com.cobblemon.mod.common.util.MovingSoundInstance
import generations.gg.generations.core.generationscore.common.world.sound.WalkmonSoundManager
import net.minecraft.client.Minecraft
import net.minecraft.sounds.SoundSource

object S2CPlaySoundHandler : com.cobblemon.mod.common.api.net.ClientNetworkPacketHandler<S2CPlaySoundPacket> {
    override fun handle(packet: S2CPlaySoundPacket, client: Minecraft) {
        val player = Minecraft.getInstance().player ?: return
        if(packet.play) {
            WalkmonSoundManager.playSound(player.uuid, packet.soundEvent, packet.length, packet.trackId)
        } else {
            WalkmonSoundManager.stopSound(player.uuid)
        }
    }
}
